sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala

```diff
@@ -156,9 +160,22 @@ case class Stack(children: Seq[Expression]) extends Generator {
     }
   }
 
+  def findDataType(column: Int): DataType = {
+    // Find the first data type except NullType.
+    val firstDataIndex = (column % numFields) + 1
+    for (i <- firstDataIndex until children.length by numFields) {
+      if (children(i).dataType != NullType) {
+        return children(i).dataType
+      }
+    }
+    // If all values of the column are NullType, use it.
+    NullType
+  }
+
   override def elementSchema: StructType =
-    StructType(children.tail.take(numFields).zipWithIndex.map {
-      case (e, index) => StructField(s"col$index", e.dataType)
+    StructType(children.tail.take(numFields).zipWithIndex.map { case (e, index) =>
```

Comment:
Do we need to change `elementSchema`? It seems to me that `StackCoercion` will convert all `NullType`, except the column are all `NullType`.